### DISCUSSION TITLE
Love ZED so far but missing eslint autofix feature is holding me back from using it

### DISCUSSION THREAD

**Discussion by DennisKraaijeveld** [zed-industries/zed#8433] [Feature Requests]
Hi!

Maybe I am doing something wrong or missing the correct config, but for instance my imports are not automatically sorted, while this is the case in VS Code. Is there something I could do or is this something which is on a roadmap? I really love using Zed and I want to try it out as my main IDE, but like how it is right now it's impossible for me to use for more than a few hours before getting frustrated.

**Comment by Moshyfawn** [comment]
No ESLint autofix at this time, but feature request is already tracked in #4325 and prioritised based on "👍".

**Comment by mrnugget** [comment]
See here: we do have this in the latest preview version:


https://github.com/zed-industries/zed/issues/4840#issuecomment-1961282033
https://github.com/zed-industries/zed/issues/4325#issuecomment-1966134129

**Reply by mrnugget** [reply]
See docs here: https://zed.dev/docs/configuring-zed#code-actions-on-format

**Reply by itsmnthn** [reply]
yes, there is support for eslint & flat config with auto fix on save. But I feel the eslint plugins rules are not being applied #9144

**Reply by DennisKraaijeveld** [reply]
Yup, exactly. It doesn't respect any of my rules in my eslint config, resulting in a wrong import order after sorting it. Or is there a way to define sorting rules for import order within the Zed config?

**Reply by mrnugget** [reply]
Can you give me a reproducible example? Eslint config and a file that should be sorted?

**Reply by DennisKraaijeveld** [reply]
Eslint Config
```cjs
const vitestFiles = ['app/**/__tests__/**/*', 'app/**/*.{spec,test}.*']
const testFiles = ['**/tests/**', ...vitestFiles]
const appFiles = ['app/**']

/** @type {import('@types/eslint').Linter.BaseConfig} */
module.exports = {
	extends: [
		'@remix-run/eslint-config',
		'@remix-run/eslint-config/node',
		'prettier',
	],
	rules: {
		'@typescript-eslint/consistent-type-imports': [
			'warn',
			{
				prefer: 'type-imports',
				disallowTypeAnnotations: true,
				fixStyle: 'inline-type-imports',
			},
		],
		'import/no-duplicates': ['warn', { 'prefer-inline': true }],
		'import/consistent-type-specifier-style': ['warn', 'prefer-inline'],
		'import/order': [
			'warn',
			{
				alphabetize: { order: 'asc', caseInsensitive: true },
				groups: [
					'builtin',
					'external',
					'internal',
					'parent',
					'sibling',
					'index',
				],
			},
		],
		'unused-imports/no-unused-imports': 'error',
	},
	overrides: [
		{
			plugins: ['remix-react-routes', 'unused-imports'],
			files: appFiles,
			excludedFiles: testFiles,
			rules: {
				'remix-react-routes/use-link-for-routes': 'error',
				'remix-react-routes/require-valid-paths': 'error',
				// disable this one because it doesn't appear to work with our
				// route convention. Someone should dig deeper into this...
				'remix-react-routes/no-relative-paths': [
					'off',
					{ allowLinksToSelf: true },
				],
				'remix-react-routes/no-urls': 'error',
				'no-restricted-imports': [
					'error',
					{
						patterns: [
							{
								group: testFiles,
								message: 'Do not import test files in app files',
							},
						],
					},
				],
			},
		},
		{
			extends: ['@remix-run/eslint-config/jest-testing-library'],
			files: vitestFiles,
			rules: {
				'testing-library/no-await-sync-events': 'off',
				'jest-dom/prefer-in-document': 'off',
			},
			// we're using vitest which has a very similar API to jest
			// (so the linting plugins work nicely), but it means we have to explicitly
			// set the jest version.
			settings: {
				jest: {
					version: 28,
				},
			},
		},
	],
}
```
Before saving in Zed
<img width="888" alt="Screenshot 2024-03-13 at 08 47 11" src="https://github.com/zed-industries/zed/assets/50783193/8b9e5730-c22d-4bd3-a467-3bacf1eb88ef">

After saving in Zed
<img width="834" alt="Screenshot 2024-03-13 at 08 47 54" src="https://github.com/zed-industries/zed/assets/50783193/64f30711-a777-42e0-abb4-efa8f2d7392e">

Saving this in VSCode
<img width="660" alt="Screenshot 2024-03-13 at 08 48 37" src="https://github.com/zed-industries/zed/assets/50783193/82affa02-c622-492a-b660-887f8646f634">

When I remove some whitespace/empty lines and move the top import (the one which is still giving a import order warning after being saved in Zed) to the bottom, and save again, than sometimes it will sort perfectly. I feel like having empty lines between imports introduce some bugs as well. For example:

```
import { SendFeedback, ThemeSwitch, useRootLoaderData } from '#app/root.tsx'
import { useOptionalUser } from '#app/utils/user.ts'

import { Link } from '@remix-run/react'

import clsx from 'clsx'

import { motion, useScroll, useTransform } from 'framer-motion'
import { forwardRef } from 'react'

import { CommandMenu } from './ui/command-menu'
import { UserNav } from './user-nav'
```

If you randomize your breaks between import, it will give you a different import order every time you save.

**Reply by mrnugget** [reply]
What's your Zed config that you use here?

**Reply by mrnugget** [reply]
@DennisKraaijeveld we now support flat ESLint configs on `main`: https://github.com/zed-industries/zed/issues/7271 I hope that this will also fix what you're running into.
